pkg/daemon/certificate_writer.go

```diff
@@ -101,92 +117,231 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
-				return err
-			}
+	oldAnno := dn.node.Annotations[constants.ControllerConfigSyncServerCA]
+	annos := map[string]string{
+		constants.ControllerConfigResourceVersionKey: controllerConfig.ObjectMeta.ResourceVersion,
+	}
+	if dn.node.Annotations[constants.ControllerConfigSyncServerCA] != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] {
+		annos[constants.ControllerConfigSyncServerCA] = controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation]
+
+	}
+	if _, err := dn.nodeWriter.SetAnnotations(annos); err != nil {
+		return fmt.Errorf("failed to set annotations on node: %w", err)
+	}
+
+	klog.Infof("Certificate was synced from controllerconfig resourceVersion %s", controllerConfig.ObjectMeta.ResourceVersion)
+
+	if controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue && oldAnno != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] && cmErr == nil {
+		if err := runCmdSync("systemctl", "stop", "kubelet"); err != nil {
+			return err
+		}
+
+		err = os.Remove("/var/lib/kubelet/kubeconfig")
+		if err != nil {
+			return fmt.Errorf("Could not remove kubelet's kubeconfig file: %v", err)
 		}
 
-		annos := map[string]string{
-			constants.ControllerConfigResourceVersionKey: controllerConfig.ObjectMeta.ResourceVersion,
+		if err := runCmdSync("systemctl", "daemon-reload"); err != nil {
+			return err
+		}
+
+		if err := runCmdSync("systemctl", "restart", "kubelet"); err != nil {
+			return err
 		}
-		if _, err := dn.nodeWriter.SetAnnotations(annos); err != nil {
-			return fmt.Errorf("failed to set ControllerConfigResourceVersion annotation on node: %w", err)
+
+		if err := wait.PollUntilContextTimeout(context.TODO(), 10*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
+			f, err := os.ReadFile("/var/lib/kubelet/kubeconfig")
+			if err != nil && os.IsNotExist(err) {
+				klog.Warningf("Failed to get kubeconfig file: %v", err)
+				return false, nil
+			} else if err != nil {
+				return false, fmt.Errorf("Unexpected error reading kubeconfig file, %v", err)
+			}
+			currentKC := clientcmdv1.Config{}
+			err = yaml.Unmarshal(f, &currentKC)
+			if err != nil {
+				return false, fmt.Errorf("Could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(f), err)
+			}
+			if !bytes.Equal(currentKC.Clusters[0].Cluster.CertificateAuthorityData, data) {
+				return false, errors.New("Cert data is not equal")
+			}
+			return true, nil
+		}); err != nil {
+			return fmt.Errorf("something went wrong while waiting for kubeconfig file to generate: %v", err)
 		}
-		klog.Infof("Certificate was synced from controllerconfig resourceVersion %s", controllerConfig.ObjectMeta.ResourceVersion)
+		klog.V(4).Infof("Finished syncing ControllerConfig %q (%v)", key, time.Since(startTime))
+		dn.CancelSIGTERM()
+		dn.Close()
+		os.Exit(0)
```

Comment:
Is this technically still necessary with the health check exit? We can technically still use the old cert until the daemon starts erroring, right?

Reply:
hmmmm. yeah probably. But, sometimes we don't error when we restart right @sergiordlr ? The health check hard kill is only if this restart does not do the job.

I can try removing this to see if it does anything? though, I thought we initially had this bc the kubelet restart necessitated a daemon restart?

Reply:
Yes, sometimes when we restart some nodes don't show any error. I don't know why, tbh. After the restart some nodes are OK and some nodes start reporting errors. The health check fixes the ones that had errors.

@yuqi-zhang's idea should work anyway, right? Nevertheless, we haven't tested this scenario.

Reply:
in the most recent commit, i did it this way. I am hoping this reduces the overall downtime of the MCO by removing 1 potential restart.